src/client/dfuse/ops/readdir.c

```diff
@@ -166,23 +170,58 @@ dfuse_cb_readdir(fuse_req_t req, struct dfuse_inode_entry *inode,
 
 	D_ASSERT(oh);
 
+	/*
+	 * the DFS size should be less than what we want in fuse to account for
+	 * the fuse metadata for each entry, so just use 1/2 for now.
+	 */
+	buf_size = size * READDIR_BLOCKS / 2;
+
 	if (offset == 0) {
 		/*
 		 * if starting from the begnning, reset the anchor attached to
 		 * the open handle.
 		 */
 		memset(&oh->doh_anchor, 0, sizeof(oh->doh_anchor));
-	} else {
+
+		/** also reset dir stream and buffer offsets */
+		oh->doh_fuse_off = 0;
+		oh->doh_cur_off = 0;
+		oh->doh_idx = 0;
+	} else if (offset != oh->doh_fuse_off) {
+		uint32_t num, keys;
+
 		/*
-		 * otherwise the offset itself should be the anchor that was
-		 * returned from the last call on the OH, otherwise the offset
-		 * was corrupted.
+		 * otherwise we are starting at an earlier offset where we left
+		 * off on last readdir, so restart by first enumerating that
+		 * many entries. This is the telldir/seekdir use case.
 		 */
-		if (offset != (off_t)&oh->doh_anchor)
-			D_GOTO(err, rc = EIO);
+
+		memset(&oh->doh_anchor, 0, sizeof(oh->doh_anchor));
```

Comment:
Is there an initialiser for this?

Reply:
in the oh allocation in opendir

Reply:
I meant a daos_anchor_init() function or similar.

Reply:
ah, no there isn't.
Liang is considering updating the interface for the anchors.